chore: bump pydantic to 2.11 and pydantic-settings to 2.9

bowenliang123 opened the pull request:
# Summary

- to close #15048
- pydantic 2.11 enhances 2x performance improvements and less memory usage since 2.9: https://pydantic.dev/articles/pydantic-v2-11-release
- bump pydantic dependencies:
```
pydantic = "~2.11.2"
pydantic-settings = "~2.9.1"
pydantic_extra_types = "~2.10.3"
```

<img width="862" alt="image" src="https://github.com/user-attachments/assets/6dc7ef82-b550-4ce9-b867-b4d5fa899453" />


> [!Tip]
> Close issue syntax: `Fixes #<issue number>` or `Resolves #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.


# Screenshots

| Before | After |
|--------|-------|
| ...    | ...   |

# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

Viicos commented on api/configs/middleware/__init__.py:
```diff
@@ -171,7 +171,7 @@ def SQLALCHEMY_DATABASE_URI(self) -> str:
     RETRIEVAL_SERVICE_EXECUTORS: NonNegativeInt = Field(
         description="Number of processes for the retrieval service, default to CPU cores.",
-        default=os.cpu_count(),
+        default_factory=lambda: NonNegativeInt(str(os.cpu_count())),
```
`NonNegativeInt` is defined as an annotated form:

```python
NonNegativeInt = Annotated[int, annotated_types.Ge(0)]
```

So in theory this shouldn't be necessary?

bowenliang123 replied:
Yes, but all the changes are minimum required changes for passing the type checks.

bowenliang123 replied:
Oh, I checked again the errors from pymy. Changed to use `default=os.cpu_count() or 4,` in this case. The violation does not come from type wrapper , but form the optional None from `os_count()` instead. Thanks!

bowenliang123 commented:
cc @laipz8200

bowenliang123 committed:
update pydantic 211

fix mypy violations

update

setting default value for RETRIEVAL_SERVICE_EXECUTORS to 1

fix default_factory with no-arg constructor to `lambda: str(uuid4())`

change to use default_factory for optional fields

nit

nit

pydantic-settings~=2.9.1

pydantic~=2.11.4

laipz8200 commented on api/configs/middleware/__init__.py:
```diff
@@ -171,7 +171,7 @@ def SQLALCHEMY_DATABASE_URI(self) -> str:
     RETRIEVAL_SERVICE_EXECUTORS: NonNegativeInt = Field(
         description="Number of processes for the retrieval service, default to CPU cores.",
-        default=os.cpu_count(),
+        default=os.cpu_count() or 8,
```
why use 8 here?

bowenliang123 replied:
The `os.cpu_count()` returns `Optional[int]` which violates `mypy` type checks for NonNegativeInt. Adding an default int to pass the checks. In general most cases , `os.cpu_count()` does return valid int and it woun't fall to the default value.

bowenliang123 replied:
Changed the default value to 1. Solved.

laipz8200 commented on api/core/plugin/entities/endpoint.py:
```diff
@@ -23,8 +23,8 @@ class EndpointProviderDeclaration(BaseModel):
 
-    settings: list[ProviderConfig] = Field(default_factory=list)
-    endpoints: Optional[list[EndpointDeclaration]] = Field(default_factory=list)
+    settings: list[ProviderConfig] = Field(default=[])
```
Can you explain this change?

bowenliang123 replied:
This fixes the `mypy` type checks.

laipz8200 commented:
I don't think changing `default_factory=list` to `default=[]` is a good idea. Even though Pydantic can handle it correctly, it is inconsistent with Python's behavior of using default values. I believe this modification is unnecessary.

To ensure the program runs correctly, if `os.cpu_count()` cannot be obtained, the default value should be set to 1.

The other parts LGTM.

bowenliang123 commented:
The 8 for defualt CPU value is referrring the similar hardcoded value in `anyio` lib. And now it's set to 1, as you suggested.

As for the changes to `default=[]`, this could be responding to `Optional[T]` type in mypy checkes. Or shall we leave `default_factory` as it is and set `default=None` in response?

bowenliang123 commented:
After checking the source code of Pydantic's code, we could see `default_factory` and `default` has the same result in our cases.  If `default_factory` method is set, it will be called with no argumenets as `default_factory()`, otherwise, it will use the `default` value.

<img width="992" alt="image" src="https://github.com/user-attachments/assets/50adc847-3e12-4a87-a82f-dd55f733c5cd" />

So it's the same for us. I'm sure that it's the right way to change `default_factory=list` to `default=[]` with no side effects and passing the mypy type checks.

bowenliang123 commented:
Alright, I have figured a better way to pass the mypy checks and satisfied your preference of keeping `default_factory`.

Original code:
```
    loop_variables: Optional[list[LoopVariableData]] = Field(default_factory=list)
```

Getting Error in checks:
```
core/workflow/nodes/loop/entities.py:29: error: Argument "default_factory" to "Field" has incompatible type "type[list[_T]]"; expected "Callable[[], Never] | Callable[[dict[str, Any]], Never]"  
```

Fixed in  default_factory with [_T]:
```
    loop_variables: Optional[list[LoopVariableData]] = Field(default_factory=list[LoopVariableData])
```

laipz8200 commented:
Using `default_factory` can pass mypy checks in the current version. If this change was introduced by an upstream update, wouldn't waiting for the upstream to fix it be a better solution?

bowenliang123 commented:
I don't think the Pydantic will regard this a blocking issue or have a fix in it, as it occurs only in mypy checkings of static types. As I'm also reported this regression via adding Dify as third-party tests in Pydantic (https://github.com/pydantic/pydantic/pull/11719). @Viicos from Pydantic team has also reviewed this PR as well.

bowenliang123 commented:
The change the type for `default_factory` in Pydantic's Field is reasonable to me. And I feel it's alright that mypy requests the implementation of default_factory align to the type definition. So just let us to decide the possible workaround for it , or just accept these minor changes as update `list` to `list[T]` for this argument.

laipz8200 commented on api/core/workflow/utils/condition/entities.py:
```diff
@@ -39,7 +39,7 @@ class SubCondition(BaseModel):
 class SubVariableCondition(BaseModel):
     logical_operator: Literal["and", "or"]
-    conditions: list[SubCondition] = Field(default=list)
+    conditions: list[SubCondition] = Field(default=[])
```
Can we also use default_factory here?

bowenliang123 replied:
Sure. Changed to use `default_factory=list` here.

bowenliang123 committed:
update uv.lock

laipz8200 reviewed (approved):
LGTM
